src/plexil-adapter/OwInterface.cpp

```diff
@@ -237,88 +237,44 @@ static void handle_joint_fault (Joint joint, int joint_index,
   handle_overtorque (joint, msg->effort[joint_index]);
 }
 
-void OwInterface::systemFaultMessageCallback
-(const  ow_faults::SystemFaults::ConstPtr& msg)
+template <typename T1, typename T2>
+void OwInterface::faultCallback (T1 msg_val, T2& fmap,
+                                 const string& component)
 {
-  // Publish all joint information for visibility to PLEXIL and handle any
-  // system-level fault messages.
-  uint64_t msg_val = msg->value;
-
-  for (auto const& entry : systemErrors){
+  for (auto const& entry : fmap) {
     string key = entry.first;
-    uint64_t value = entry.second.first;
-    bool b = entry.second.second;
-
-    if (checkFaultMessages("SYSTEM", msg_val, key, value, b)) {
-      systemErrors[key].second = !systemErrors[key].second;
+    T1 value = entry.second.first;
+    bool exists = entry.second.second;
+    bool faulty = msg_val & value == value;
+    if (!exists && faulty) {
+      ROS_WARN ("Fault in %s: %s", component.c_str(), key.c_str());
     }
-
+    else if (exists && !faulty) {
+      ROS_WARN ("Resolved fault in %s: %s", component.c_str(), key.c_str());
+    }
+    if (exists != faulty) fmap[key].second = !fmap[key].second;
```

Comment:
I personally do understand what the code is doing here but I still think it hard to read,  
I suggest that you refactor the code as follows:
```C++
   if (!exists && faulty) {
      ROS_WARN ("Fault in %s: %s", component.c_str(), key.c_str());
      fmpa[key].second = true;
    }
    else if (exists && !faulty) {
      ROS_WARN ("Resolved fault in %s: %s", component.c_str(), key.c_str());
      fmap[key].second = false;
    }
``` 

This is more readable and slightly more efficient :rocket:

Reply:
Agreed, changed.